command/agent/cache/cacheboltdb/bolt.go

```diff
@@ -0,0 +1,330 @@
+
+	// SecretLeaseType - Bucket/type for leases with secret info
+	SecretLeaseType = "secret-lease"
+
+	// AuthLeaseType - Bucket/type for leases with auth info
+	AuthLeaseType = "auth-lease"
+
+	// TokenType - Bucket/type for auto-auth tokens
+	TokenType = "token"
+
+	// AutoAuthToken - key for the latest auto-auth token
+	AutoAuthToken = "auto-auth-token"
+
+	// RetrievalTokenBucket names the bucket that stores the retrieval material
+	// needed for unlocking (decrypting) the persisted tokens and leases
+	RetrievalTokenBucket = "retrieval-token"
+
+	// RetrievalTokenMaterial is the actual key or token in the key bucket
+	RetrievalTokenMaterial = "retrieval-token-material"
+)
+
+// BoltStorage is a persistent cache using a bolt db. Items are organized with
+// the encryption key as the top-level bucket, and then leases and tokens are
+// stored in sub buckets.
+type BoltStorage struct {
+	db         *bolt.DB
+	rootBucket string
+	logger     hclog.Logger
+	keyMgr     crypto.KeyManager
+	aad        string
+}
+
+// BoltStorageConfig is the collection of input parameters for setting up bolt
+// storage
+type BoltStorageConfig struct {
+	Path       string
+	Logger     hclog.Logger
+	AAD        string
+	KeyManager crypto.KeyManager
+}
+
+// NewBoltStorage opens a new bolt db at the specified file path and returns it.
+// If the db already exists the buckets will just be created if they don't
+// exist.
+func NewBoltStorage(config *BoltStorageConfig) (*BoltStorage, error) {
+	dbPath := filepath.Join(config.Path, DatabaseFileName)
+	db, err := bolt.Open(dbPath, 0600, &bolt.Options{Timeout: 1 * time.Second})
+	if err != nil {
+		return nil, err
+	}
+	err = db.Update(func(tx *bolt.Tx) error {
+		return createBoltSchema(tx)
+	})
+	if err != nil {
+		return nil, err
+	}
+	bs := &BoltStorage{
+		db:         db,
+		rootBucket: rootBucketName,
+		logger:     config.Logger,
+		keyMgr:     config.KeyManager,
+		aad:        config.AAD,
+	}
+	if config.KeyManager != nil {
+		bs.SetKeyMgr(config.KeyManager)
+	}
+	return bs, nil
+}
+
+func createBoltSchema(tx *bolt.Tx) error {
+	_, err := tx.CreateBucketIfNotExists([]byte(RetrievalTokenBucket))
+	if err != nil {
+		return fmt.Errorf("failed to create key bucket: %w", err)
+	}
+	root, err := tx.CreateBucketIfNotExists([]byte(rootBucketName))
+	if err != nil {
+		return fmt.Errorf("failed to create bucket %s: %w", rootBucketName, err)
+	}
+	_, err = root.CreateBucketIfNotExists([]byte(TokenType))
+	if err != nil {
+		return fmt.Errorf("failed to create token sub-bucket: %w", err)
+	}
+	_, err = root.CreateBucketIfNotExists([]byte(AuthLeaseType))
+	if err != nil {
+		return fmt.Errorf("failed to create auth lease sub-bucket: %w", err)
+	}
+	_, err = root.CreateBucketIfNotExists([]byte(SecretLeaseType))
+	if err != nil {
+		return fmt.Errorf("failed to create secret lease sub-bucket: %w", err)
+	}
+
+	// check and set file version in the root bucket
+	version := root.Get([]byte(storageVersionKey))
```

Comment:
Can we do this first and bail early on error?

Reply:
Yep, changed in https://github.com/hashicorp/vault/pull/10938/commits/0dd3ae4cdddefe9fa567b29be9c25fbd3a51fb93